VariantShredding.md

```diff
@@ -25,290 +25,316 @@
+| float                       | FLOAT                        |
+| double                      | DOUBLE                       |
+| decimal4                    | DECIMAL(precision, scale)    |
+| decimal8                    | DECIMAL(precision, scale)    |
+| decimal16                   | DECIMAL(precision, scale)    |
+| date                        | DATE                         |
+| timestamp                   | TIMESTAMP(true, MICROS)      |
+| timestamp without time zone | TIMESTAMP(false, MICROS)     |
+| binary                      | BINARY                       |
+| string                      | STRING                       |
+| array                       | LIST; see Arrays below       |
+| object                      | GROUP; see Objects below     |
 
-# Unshredded values
+#### Primitive Types
 
-If all values can be represented at a given level by whichever of `object`, `array`, or `typed_value` is present, `variant_value` is set to null.
+Primitive values can be shredded using the equivalent Parquet primitive type from the table above for `typed_value`.
 
-If a value cannot be represented by whichever of `object`, `array`, or `typed_value` is present in the schema, then it is stored in `variant_value`, and the other fields are set to null.
-In the Parquet example above, if field `a` was an object or array, or a non-integer scalar, it would be stored in `variant_value`.
+Unless the value is shredded as an object (see [Objects](#objects)), `typed_value` or `value` (but not both) must be non-null.
 
-If a value is an object, and the `object` field is present but does not contain all of the fields in the value, then any remaining fields are stored in an object in `variant_value`.
-In the Parquet example above, if field `b` was an object of the form `{"c": 1, "d": 2}"`, then the object `{"d": 2}` would be stored in `variant_value`, and the `c` field would be shredded recursively under `object.c`.
+#### Arrays
 
-Note that an array is always fully shredded if there is an `array` field, so the above consideration for `object` is not relevant for arrays: only one of `array` or `variant_value` may be non-null at a given level.
+Arrays can be shredded using a 3-level Parquet list for `typed_value`.
 
-# Using variant_value vs. typed_value
+If the value is not an array, `typed_value` must be null.
+If the value is an array, `value` must be null.
 
-In general, it is desirable to store values in the `typed_value` field rather than the `variant_value` whenever possible.
-This will typically improve encoding efficiency, and allow the use of Parquet statistics to filter at the row group or page level.
-In the best case, the `variant_value` fields are all null and the engine does not need to read them (or it can omit them from the schema on write entirely).
-There are two main motivations for including the `variant_value` column:
+The list `element` must be a required group that contains `value` and `typed_value` fields.
```

Comment:
I wonder if it might be easier to rephrase this logic overall in terms of rules to value and typed value presence for VARIANT annotated types?  Something off the top of my head but would another logical type VARIANT_ELEMENT help make things clearer?

Reply:
On further thinking a separate annotation probably does not add value here